subsys/instrumentation/common/instr_common.c

```diff
@@ -0,0 +1,565 @@
+		}
+		for (int k = 0; k < sizeof(disco_func[i].delta_t); k++) {
+			uart_poll_out(uart_dev, *((uint8_t *)&disco_func[i].delta_t + k));
+		}
+	}
+
+	printk("-*-!\n"); // Terminator
+}
+
+__no_instrumentation__
+void push_callee_timestamp(void *callee)
+{
+	int i, found = 0;
+
+	for (i = 0; i < disco_func_pos; i++) {
+		if (disco_func[i].func_addr == callee) {
+			found = 1;
+			break;
+		}
+	}
+
+	if (!found) {
+		if (disco_func_pos >= NUM_DISCO_FUNCS)
+			return; // No more space to add another func
+
+		disco_func[disco_func_pos].func_addr = callee;
+		disco_func[disco_func_pos].time_stack_pos = 0;
+		disco_func[disco_func_pos].delta_t = 0;
+		i = disco_func_pos;
+		disco_func_pos++;
+	}
+
+	// Update time stack for function at 'func' address
+	if (disco_func[i].time_stack_pos < 16) {
+		// TODO(gromero): probably there can be only one stack tho,
+		// instead of one per function, even for inlined ones, this will
+		// allow squeezing a bit more the memory layout by avoiding
+		// "holes" in it.
+		// push time to func's time stack
+		int p = disco_func[i].time_stack_pos;
+		disco_func[i].time_stack[p] = instr_timestamp();
+		disco_func[i].time_stack_pos++;
+	}
+}
+
+__no_instrumentation__
+void pop_callee_timestamp(void *callee)
+{
+	uint32_t dt;
+
+	for (int i = 0; i < disco_func_pos; i++) {
+		if (disco_func[i].func_addr == callee) {
+			int p = --disco_func[i].time_stack_pos;
+
+			// delta = exit timestamp - entry timestamp
+			dt = instr_timestamp() - disco_func[i].time_stack[p];
+			disco_func[i].delta_t += dt;
+
+			return;
+		}
+	}
+
+	// Track number of unbalanced entries
+	unbalanced++;
+}
+
+__no_instrumentation__
+int is_context_available(void) {
+	int save_context = 0;
+
+	IF_ENABLED(CONFIG_INSTRUMENTATION_SAVE_CONTEXT, (save_context = 1));
+
+	return save_context;
+}
+
+__no_instrumentation__
+void save_context(struct instr_record* record) {
+	k_tid_t curr_thread;
+
+	curr_thread = k_current_get();
+
+	record->context.cpu = arch_proc_id();
+	record->context.thread_id = curr_thread;
+	record->context.mode = curr_thread ? k_thread_priority_get(curr_thread): 0;
+
+	/*
+	 * k_thread_name_copy is not used here because this function and strncpy
+	 * would have to be added to the compiler exclude list. However, both
+	 * functions are important to instrument for trace/profile, hence below
+	 * we don't use any function to set the thread name.
+	 */
+	if (curr_thread) {
+		int i = 0;
+		char *thread_name = curr_thread->name;
+		while (*thread_name != '\0') {
+			record->context.thread_name[i] = *thread_name;
+			i++;
+			thread_name++;
+		}
+		record->context.thread_name[i] = '\0';
+	} else {
+		int i;
+		char thread_name[] = "thread-none";
+		for (i = 0; thread_name[i] != '\0'; i++) {
+			record->context.thread_name[i] = thread_name[i];
+		}
+		record->context.thread_name[i] = '\0';
+	}
+}
+
+__no_instrumentation__
+void instr_event_handler(enum instr_event_types type, void *callee, void *caller)
```

Comment:
This doesn't seem to account for SMP in any way. Presumably it would be some metadata in the event *or* an entirely different stream (probably preferable to avoid locking) per core.

Reply:
The only downside of this is that output streams need to be muxed in some way then, and will have some lock likely around it.